ludwig/hyperopt/utils.py

```diff
@@ -51,3 +54,90 @@ def should_tune_preprocessing(config):
         if f"{PREPROCESSING}." in param_name:
             return True
     return False
+
+
+def parameter_to_dict(name, value):
+    if name == ".":
+        # Parameter name ".", means top-level config
+        return value
+
+    parameter_dict = {}
+    curr_dict = parameter_dict
+    name_list = name.split(".")
+    for i, name_elem in enumerate(name_list):
+        if i == len(name_list) - 1:
+            curr_dict[name_elem] = value
+        else:
+            name_dict = curr_dict.get(name_elem, {})
+            curr_dict[name_elem] = name_dict
+            curr_dict = name_dict
+    return parameter_dict
+
+
+def feature_list_to_dict(config: Dict[str, Any]) -> Dict[str, Any]:
+    input_features_dict = {}
+    for feature in config[INPUT_FEATURES]:
+        input_features_dict[feature[NAME]] = feature
+
+    output_features_dict = {}
+    for feature in config[OUTPUT_FEATURES]:
+        output_features_dict[feature[NAME]] = feature
+
+    config = copy.copy(config)
+    config[INPUT_FEATURES] = input_features_dict
+    config[OUTPUT_FEATURES] = output_features_dict
+    return config
+
+
+def feature_dict_to_list(config: Dict[str, Any]) -> Dict[str, Any]:
+    # This works because Python dicts are order-preserving, so we do not need to
+    # do anything special to map from a key in the dict to an index in a list
+    input_features_list = []
+    for feature in config[INPUT_FEATURES].values():
+        input_features_list.append(feature)
+
+    output_features_list = []
+    for feature in config[OUTPUT_FEATURES].values():
+        output_features_list.append(feature)
+
+    config = copy.copy(config)
+    config[INPUT_FEATURES] = input_features_list
+    config[OUTPUT_FEATURES] = output_features_list
+    return config
+
+
+def substitute_parameters(
+    config: Dict[str, Any],
+    parameters: Dict[str, Any],
+):
+    """Update Ludwig config with parameters sampled from the Hyperopt sampler."""
+
+    # Collect the sets of names for each feature grouping so we can map feature names to
+    # groups
+    input_feature_names = {feature[NAME] for feature in config[INPUT_FEATURES]}
+    output_feature_names = {feature[NAME] for feature in config[OUTPUT_FEATURES]}
+
+    # Features in the user config are provided as a list, but in hyperopt we reference
+    # features by name, so convert temporarily to a dict to simplify the mergep process.
+    config = feature_list_to_dict(config)
+
+    # Merge parameters into the user configuration in order. As such, if there are conflicting
+    # params, the later params will take precedence.
+    for name, value in parameters.items():
+        # User params are provided as <feature_name>.<param>, but we group input / output features
+        # together during the merge to make it easier and unambiguous to convert back and forth
+        # TODO(travis): we should revisit the user format here, as it silently breaks situations
+        # where the user has a feature named "trainer", "combiner", etc.
```

Comment:
This is a quick check to see if the top-level config section keywords of Ludwig are within the feature names from lines 117-118 right? Would the ideal outcome be to just raise an error and ask the user to rename their feature? 

I think that users shouldn't be allowed to have feature names that have any of the "special" keywords that we use within the Ludwig config.

Reply:
I would probably want to change the way we reference features in hyperopt to include `input_features.` in front to resolve ambiguity. But it's a big breaking change, so probably not worth doing just now. I agree a check in schema validation would be useful to catch this as early as possible.

Reply:
Got it! Might be a good change to include in 0.7 along with this change.